18405 duw v2 review update

chriskim2311 opened the pull request:
## Are you removing, renaming or moving a folder in this PR?
- [x] No, I'm not changing any folders (skip to Summary and delete the rest of this section)
- [ ] Yes, I'm removing, renaming or moving a folder

Did you change site-wide styles, platform utilities or other infrastructure?
- [x] No
- [ ] Yes, and I used the [proxy-rewrite steps](https://github.com/department-of-veterans-affairs/vets-website/tree/main/src/applications/proxy-rewrite#that-sounds-normal-so-whats-the-proxy-all-about) to test the injected header scenario

## Summary
This PR changes the behavior of the review page of v2 DUW. 

Changes include: 
Always continuing the flow of questions when using navigation buttons.
Added edit mode to return to the review screen when editing a question without a forked flow.
Added 2 expandable alerts: when editing a forkable question and returning to the edit screen. 


## Related issue(s)
https://github.com/department-of-veterans-affairs/va.gov-cms/issues/18405

## Testing done
Mural of flows: https://app.mural.co/t/departmentofveteransaffairs9999/m/departmentofveteransaffairs9999/1696352911500/80253b1c7200212cbcbcfb940d2cf4a29fed8417?sender=ua4eca0cc168eff677bd03940

Figma for design review updates: https://www.figma.com/design/XHQAQ8ZDKKsDcXfBn5ulhh/Discharge-Upgrade-Wizard?node-id=379-5837&t=OGAz4YMd7SiRQFLU-0

Tested back and continue button functionality works as intended and without clicking edit you will continue through each question in the flow.

Tested on edit if a question was changed for a non-forking question you go back to the review page. If you change the answer for a question that forks you will see the new questions of the flow.

5 questions to test for this:
  DISCHARGE_YEAR,
  REASON,
  PREV_APPLICATION,
  PREV_APPLICATION_TYPE,
  PREV_APPLICATION_YEAR,

## Screenshots

Review screen now has question labels instead of just responses.
<img width="1022" alt="Cursor_and_How_To_Apply_For_A_Discharge_Upgrade___Veterans_Affairs" src="https://github.com/user-attachments/assets/dd0ed5d0-7296-46c1-826b-163b190ff10f">

Clicking edit and going back to a question that can lead to a different question flow
<img width="1066" alt="Cursor_and_How_To_Apply_For_A_Discharge_Upgrade___Veterans_Affairs" src="https://github.com/user-attachments/assets/5fbc84cf-92ae-4bea-83b4-1d84f71c35b3">

Continuing through the new flow and see new alert about the questions in review being different
<img width="1059" alt="How_To_Apply_For_A_Discharge_Upgrade___Veterans_Affairs_and_page-navigation_js_—_Code_Repos" src="https://github.com/user-attachments/assets/4798e478-12f8-46e3-a921-bd2d26738296">


## What areas of the site does it impact?
DUW staging v2

## Acceptance criteria

### Quality Assurance & Testing

- [ ] I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [ ] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [ ] Linting warnings have been addressed
- [ ] Documentation has been updated ([link to documentation](#) \*if necessary)
- [x] Screenshot of the developed feature is added
- [ ] [Accessibility testing](https://depo-platform-documentation.scrollhelp.site/developer-docs/wcag-2-1-success-criteria-and-foundational-testing) has been performed

### Error Handling

- [x] Browser console contains no warnings or errors.
- [ ] Events are being sent to the appropriate logging solution
- [ ] Feature/bug has a monitor built into Datadog or Grafana (if applicable)

Commit:
updates wip

Commit:
Merge branch 'main' of https://github.com/department-of-veterans-affairs/vets-website into 18405-duw-v2-review-update

Commit:
updates for review screen changes

Commit:
updates

Commit:
comment out test

randimays commented:
Reminder to delete this section of the PR description if it doesn't apply, since the description can get pretty cluttered.

<img width="824" alt="Screenshot 2024-08-12 at 3 43 07 PM" src="https://github.com/user-attachments/assets/ed04e735-a031-4b61-8d2d-d023db3ffa41">

laflannery commented:
@chriskim2311 a few comments:
1. I just realized that the questions are currently using the `hint` prop but shouldn't be. They need to use the form description slot, like on the PACT Act form. [This is the component pattern](https://design.va.gov/storybook/?path=/docs/uswds-va-radio--docs#forms-pattern-single) that should be followed.
2. Should the form screens have a max width? It all seems much wider than Income Limits and PACT Act and it might make sense to have them all be uniform?
3. On the review screen the "Edit" links have no `href` prop which makes them not actually links and therefore not available to the keyboard or assistive tech. They have to have an `href` value. On PACT Act they all just go to "#", would that work here as well?

randimays commented on src/applications/discharge-wizard/components/v2/ReviewPage.jsx:
```diff
@@ -37,22 +48,31 @@ const ReviewPage = ({ formResponses, router, viewedIntroPage }) => {
-              text="Edit"
-            />
+            <div className="answer-review-label">
+              <div
```
We should be using semantic markup here, perhaps a `<p>` tag instead of a `<div>` to contain this text.

randimays commented on src/applications/discharge-wizard/components/v2/ReviewPage.jsx:
```diff
@@ -37,22 +48,31 @@ const ReviewPage = ({ formResponses, router, viewedIntroPage }) => {
+                {reviewLabel}
+              </div>
+              <va-link
+                class="hydrated vads-u-padding-left--2"
```
Copy-pasta? I don't know if we need to add the `hydrated` class here.

randimays commented on src/applications/discharge-wizard/components/v2/ReviewPage.jsx:
```diff
@@ -37,22 +48,31 @@ const ReviewPage = ({ formResponses, router, viewedIntroPage }) => {
+                text="Edit"
+              />
+            </div>
+            <div>{reviewAnswer}</div>
```
Semantic markup here as well please.

randimays commented on src/applications/discharge-wizard/constants/question-data-map.js:
```diff
@@ -93,3 +93,18 @@ export const RESPONSES = Object.freeze({
+export const REVIEW_LABEL_MAP = Object.freeze({
+  SERVICE_BRANCH: 'Branch of service:',
+  DISCHARGE_YEAR: 'Year of discharge:',
+  DISCHARGE_MONTH: 'Month of discharge',
```
This one's missing the `:`

randimays commented on src/applications/discharge-wizard/tests/v2/review/ReviewPage.unit.spec.js:
```diff
@@ -0,0 +1,80 @@
+// import React from 'react';
```
Will this test be fixed in a future PR?

randimays replied:
Bumping this comment

randimays reviewed (changes requested):
I went through and tested some flows and have some observations.

<details><summary>1</summary>
<br/>
When I go through the flow normally and trigger the "What month were you discharged?" question, I see this console error:

<img width="484" alt="Screenshot 2024-08-13 at 9 10 03 AM" src="https://github.com/user-attachments/assets/69829f66-1034-4378-a651-f42f2daf32f2">

</details>

<details><summary>2</summary>
<br/>
Scenario:

- go through the flow normally with the below answers
- click Edit for 2008
- change that answer to 2009

I don't see the "Changing your answer..." warning (since the month question will be shown). The month question does display after clicking Continue, though.

<img width="600" alt="Screenshot 2024-08-13 at 9 11 16 AM" src="https://github.com/user-attachments/assets/6a6421b6-b574-4261-9ba6-84da98ff69f8">
</details>

<details><summary>3</summary>
<br/>
The month question should be converted to the human-friendly month names.

<img width="600" alt="Screenshot 2024-08-13 at 9 13 26 AM" src="https://github.com/user-attachments/assets/5a708c89-6311-4aef-ad2c-b0be6dc125ad">
</details>
<details><summary>4</summary>
<br/>
Scenario: 

- go through the flow with the below answers
- click Edit on "Tell us why you want to change your discharge paperwork."
- select the "My discharge is unjust..." reason

<img width="600" alt="Screenshot 2024-08-13 at 9 17 30 AM" src="https://github.com/user-attachments/assets/89fb96a3-5141-444e-8846-599cd368eb34">

I'm expecting to get all of the same questions after the reason question because both reasons (my original choice and the new one) are in the "Main flow" in the Mural. 

<img width="600" alt="Screenshot 2024-08-13 at 9 29 01 AM" src="https://github.com/user-attachments/assets/a5af2dc6-4a7b-4ebf-960e-285dfc9f3a8a">

When I click select another one of the "Main flow" reasons, shouldn't I go back to the Review screen? When I click Continue, I'm shown the questions I already answered before but my answers are cleared out. 

Is it expected that if you change an answer earlier in the flow, **all** questions, even ones you've already answered, will be blank going forward?

In the video below I went through the scenario I described. When I got to the question I'd already answered and noticed my answer was cleared out, I clicked Back a few times and ended up at the Review screen where only part of my answers were saved. When I clicked Continue there, I got a broken Results page (understandably since answers are missing).

https://github.com/user-attachments/assets/11349a62-3a01-41d9-99f0-66054903e00a

Note also that there's a warning about using `data-testId` versus `data-testid` (all lowercase). The lowercase version is the accepted syntax; I think it'll still work if you have camel case, but it has a console warning if you use camel case.

</details>

<details><summary>5</summary>
<br/>
When I go through any flow, and click Edit on the discharge year answer, should I be seeing the "Changing your answer..." warning? It is possible to get a different question if I select 2008 (15 years ago). I have noticed in other situations that we are showing the "Changing your answer..." where the questions afterward may not change (but we're just covering our bases), such as on the reason question.
</details>

<details><summary>6</summary>
<br/>
When I go through the flow with the below answers and click Edit on the previous application type question, there are no scenarios where the questions after that one will change (per this flow on the Mural). But I do see the "Changing your answer..." warning anyway.
<img width="600" alt="Screenshot 2024-08-13 at 10 00 51 AM" src="https://github.com/user-attachments/assets/3287bdf9-0561-41c4-b914-ed096ae520d7">

<img width="600" alt="Screenshot 2024-08-13 at 10 00 57 AM" src="https://github.com/user-attachments/assets/7ac8e767-3baa-4468-bc21-5312d65002c0">
</details>

<details><summary>7</summary>
<br/>
I see some lint warnings about variables missing in prop type declarations. We should add those if possible.
</details>

thejordanwood commented:
**Design notes:**

- I agree with Laura that this needs a max width like Income Limits and PACT Act. 
- I think there should be padding/space between the edit button and the body copy. When the copy goes under the edit button it looks a bit cluttered. 
<img width="1001" alt="Screenshot 2024-08-13 at 1 07 16 PM" src="https://github.com/user-attachments/assets/b00f9f92-620b-4399-b4ca-b8d8c2b283d2">


**Content notes:**

- When I went back to edit the DISCHARGE_YEAR question, I noticed the year of discharge now says "Jan 2009". I think this should say the year only. Randi has already noted that the month question should be converted to the human-friendly month names.
<img width="1041" alt="Screenshot 2024-08-13 at 1 00 42 PM" src="https://github.com/user-attachments/assets/219d1b2d-b7d3-44c2-a7db-c188983dc439">

Commit:
updates

Commit:
changes

chriskim2311 commented:
@laflannery Thanks! I have addressed all your comments.

@randimays Thanks for the review! Answers to your questions below:
1: Still looking into this, error is also seen on prod.

2 & 3: Changed Discharge year to show the alert and fixed review for month to show friendly name

4: When changing reason for example I have cleared all answers that have followed previously. The next question may be the same for a few answers however there are certain questions that come later in the flow that show only to specific answers to the reason question. For this reason I clear all the answers. This goes for all forked questions minus Discharge year which only clears Discharge Month.

5: For Discharge Year it clears Discharge Month, no other questions rely on this answer.

6: Previous Application Type is a tricky one here is the way to show the last question Failure to Exhaust and Prior Service that is cleared or potentially shown based on answers.
![How_To_Apply_For_A_Discharge_Upgrade___Veterans_Affairs](https://github.com/user-attachments/assets/44068c18-083c-4c50-a6da-8ef74e588d8d)

@thejordanwood Thanks! The branching questions on answer change should take your through a potential new flow of questions. For example changing discharge year from 2024 to 2009 it will take you to discharge month and continues through the flow. Prev application type is tricky, for that question and the answers to other questions in the form that match the flow take you to failure to exhaust and/or prior service the rest will go to the review screen, so that behavior you mentioned is expected.

Commit:
update back functionality

laflannery commented:
@chriskim2311 I have some additional markup changes to the review screen. Currently the way it is now, SRs hear Question, Edit, answer chosen. But it would make more sense to do this like PACT Act, Question, answer, edit. In this way they will have better context to if they actually want/need to edit their answers.  This markup also removes some of the padding to align the text with the borders and such. In the screen shot you can see the first row with the new alignment and the second with the existing. @thejordanwood Can you review and let us know if this looks better and if we should move forward with all the padding adjustments as well?

![image](https://github.com/user-attachments/assets/c373c1e2-55ed-4b81-ac5d-b5e7e138f2b5)

Updated markup: 
```
<ul class="vads-u-padding--0">
  <li class="vads-u-padding-y--1p5 vads-u-margin-bottom--0 answer-review-box">
    <div class="vads-u-display--flex vads-u-justify-content--space-between vads-u-align-items--center">
      <p>
        <span class="vads-u-font-weight--bold">Branch of service:</span>
        <br>
        <span>I served in the Navy.</span>
      </p>
      <va-link class="hydrated" href="#" name="SERVICE_BRANCH" label="Edit Branch of service:" text="Edit"></va-link>
    </div>
  </li>
And then the other <li> items go here
</ul>

laflannery commented:
2 other items:
1. After I edit an answer, I cannot use the back button on the Review screen. Example:
- I edited my year of discharge
  - It was 2022 and I changed to 2020
  - I saw the warning on the Discharge screen when I clicked edit
  - I successfully changed my year of discharge
  - **I cannot click back when I arrived back on the Review screen.**
  - I am able to click continue though
2. In slack it was mentioned that we would change "will" to "may" on the warning on the question screens. I didn't see that but I wasn't sure if it was supposed to yet so I just wanted to mention that just in case.

Commit:
update tests

Commit:
updates to functionality

Commit:
update markup

chriskim2311 commented:
@thejordanwood @laflannery padding changes in view along with the previous changes Jordan mentioned with the answer copy. 
![How_To_Apply_For_A_Discharge_Upgrade___Veterans_Affairs](https://github.com/user-attachments/assets/ed5e64f0-bd0b-4295-96d1-ea8bb72af618)

Commit:
small bug fix

thejordanwood commented:
I just reviewed again and have a few more notes. 

**Design/copy notes:**

- Review screen design looks good! cc: @laflannery 
- Can the body copy also be changed to also say "may"? It should read: If you change your answer to this question, you may be asked for more information to ensure that we provide you with the best results.
<img width="658" alt="Screenshot 2024-08-14 at 5 00 56 PM" src="https://github.com/user-attachments/assets/befad90e-5340-4f0b-b85f-c246dd1320d8">


**Flow notes:**

Expected edit flow seems good. I was testing out this scenario:

> AND Clicking back on the first new question should go back to the previous question they originally chose to edit
> 

These were my steps:

1. Editing PREV_APPLICATION from No to Yes. 
2. Clicking Continue to PREV_APPLICATION_YEAR
3. Not selecting an answer, but clicking Back instead. 
4. This took me back to the review screen. Ideally, it would take me back to the PREV_APPLICATION question. 

<img width="754" alt="Screenshot 2024-08-14 at 5 22 32 PM" src="https://github.com/user-attachments/assets/e0eb5007-3472-4528-8afe-5a34a19f294c">

randimays commented on src/applications/discharge-wizard/components/v2/ReviewPage.jsx:
```diff
@@ -4,12 +4,22 @@ import { connect } from 'react-redux';
 import { VaButtonPair } from '@department-of-veterans-affairs/component-library/dist/react-bindings';
```
There are a few lint warnings about needing to add variables to props validation. I'd recommend adding these if you can.

randimays commented on src/applications/discharge-wizard/components/v2/ReviewPage.jsx:
```diff
@@ -29,30 +39,49 @@ const ReviewPage = ({ formResponses, router, viewedIntroPage }) => {
   };
 
+  const onBackClick = () => {
+    const lastRoute = routeMap[routeMap.length - 2];
```
Consider: should we use `previous` here instead of `last` since `last` can also indicate the (not in edit mode) last question in the flow before the review screen?

randimays commented on src/applications/discharge-wizard/components/v2/questions/shared/Dropdown.jsx:
```diff
@@ -7,12 +7,18 @@ import {
 } from '@department-of-veterans-affairs/component-library/dist/react-bindings';
```
Some props validation lint warnings here that could be addressed too.

randimays commented on src/applications/discharge-wizard/reducers/v2/discharge-upgrade-wizard.js:
```diff
@@ -34,9 +37,11 @@ const {
   viewedIntroPage: false,
+  editMode: false,
+  questionFlowChanged: false,
+  routeMap: ['introduction', 'service-branch'],
```
Should these be constants from somewhere?

randimays commented on src/applications/discharge-wizard/tests/v2/display-logic.unit.spec.js:
```diff
@@ -8,13 +8,21 @@ describe('utilities: display logic', () => {
   describe('navigateForward', () => {
     describe('routing to discharge year', () => {
       const formResponses = { SERVICE_BRANCH: RESPONSES.ARMY };
+      const routeMap = ['introduction', 'service-branch'];
```
There are a lot of strings for routes in this file; should we be using constants in case those get updated?

randimays reviewed (changes requested):
Just wrapped up testing and I have a few more observations.

<details><summary>1</summary>
<br/>

Months should be the full word rather than the abbreviation

<img width="400" alt="Screenshot 2024-08-15 at 7 55 59 AM" src="https://github.com/user-attachments/assets/a321f67e-7183-41a0-ab46-d1f96979410e">

</details>

<details><summary>2</summary>
<br/>

Scenario:

- go through the flow normally with the below answers
- click Edit for 2008
- change that answer to 2009

When you change the year to 2009, then answer the month question, I think the Continue button should go back to the review screen since none of the other questions are new (and the answers are prefilled when I go through them). 

  > EXCEPT Clicking continue on the last new question should jump the user to the review page. (Assuming there are situations where new questions are perfectly encapsulated in the middle of the flow)

<img width="597" alt="Screenshot 2024-08-15 at 8 13 24 AM" src="https://github.com/user-attachments/assets/94910b54-217f-45b9-a13d-63e1ed638b05">

</details>

<details><summary>3</summary>
<br/>

Scenario:

- go through the flow with the below answers
- click Edit on "Tell us why you want to change your discharge paperwork."
- select the "My discharge is unjust..." reason

The "My discharge is unjust..." reason has the same question flow as the PTSD flow; the answers should be prefilled on this flow as they are in #2 in this comment, right? And if so, should we skip directly to the review screen after changing the reason?

<img width="604" alt="Screenshot 2024-08-15 at 8 17 38 AM" src="https://github.com/user-attachments/assets/d1511d0c-f1bc-4399-ba0f-502231362810">

</details>

<details><summary>4</summary>
<br/>

I don't think this is in ACs anywhere, but if you store a "roadmap" of sorts somewhere, is it possible to **not** show the "Changing your answers may..." warning if the Edit link is clicked on the very last question in a flow? For instance, if the flow is A -> B -> C, and the Edit link is clicked on C, and there's never a possibility of a question D, should we show the "Changing your answers may..." warning?

Likely need @thejordanwood's input on this one; I may be overthinking this.

</details>

<details><summary>5</summary>
<br/>

I think there are some questions missing in the flow below. If you say "Yes" to the previous application question, there are 1-3 other questions that should be asked but they don't show up. Same if you had previously said "No" to this question and you edit it to "Yes."

<img width="500" alt="Screenshot 2024-08-15 at 8 32 08 AM" src="https://github.com/user-attachments/assets/905af25d-d0bd-4ab9-b62f-ca01f904c25a">

Missing question(s)?
<img width="600" alt="Screenshot 2024-08-15 at 8 33 00 AM" src="https://github.com/user-attachments/assets/5fe926c4-8668-4769-91f1-6141642cbb74">

<details><summary>6</summary>
<br/>

I also noticed what Laura pointed out [in this comment](https://github.com/department-of-veterans-affairs/vets-website/pull/31319#issuecomment-2288755437): the Back button doesn't work on the review screen after a question has been edited.

chriskim2311 commented:
@randimays 
1: Done!
2: Done!
3: The flow does look similar but for unjust it will skip the prev app year question if you answer yes to the prev app question. There are a lot of one off question flows further down the question tree. Especially for Reason which is why I clear all answers after the change in forking questions. 
4: The last forking question isn't the last question but could be if you have specific answers in the form and change an answer for Prev App Type. I think the "Changing your answers may..." copy works for this instead of adding more complexity.  @thejordanwood Your input would be great on this!
5: Done! Found a bug with the prior service display conditions
6: Done!

Commit:
updates based on PR comments

thejordanwood commented:
@chriskim2311 @randimays Let's leave the wording as is for now. Dave made [a comment in Slack](https://dsva.slack.com/archives/C52CL1PKQ/p1723583747749899?thread_ts=1723494965.135929&cid=C52CL1PKQ) that we can test this in usability testing and adjust if needed.

randimays commented:
@chriskim2311 
> The flow does look similar but for unjust it will skip the prev app year question if you answer yes to the prev app question. There are a lot of one off question flows further down the question tree. Especially for Reason which is why I clear all answers after the change in forking questions.

I don't think it should be skipping the prev app year question? I'm looking at this section of the Mural and don't see anything that indicates the prev app year question should only show for certain flows. I see that prod has the same behavior; is it a mistake in the Mural? @thejordanwood 

<img width="828" alt="Screenshot 2024-08-16 at 10 19 43 AM" src="https://github.com/user-attachments/assets/9429717d-a5d5-4e02-b119-e0e97c58fb3c">

randimays commented:
@chriskim2311 Are you still working on the commented-out unit test?

Commit:
update unit test

laflannery commented:
@chriskim2311 A couple of my previous comments look like they may have been missed:
- # 1 from t[his comment here](https://github.com/department-of-veterans-affairs/vets-website/pull/31319#issuecomment-2288755437), I am still unable to use the back button in some cases
- The markup has not been adjusted to match what was in [this comment here](https://github.com/department-of-veterans-affairs/vets-website/pull/31319#issuecomment-2288735840). Padding was adjusted but the main issue here is that the order of the text within the DOM needs to be adjusted so that the Edit links are last within each list item

laflannery commented:
Note for future reference: We have limited the scope of this PR to just the review screen, not the logic and flow of edit questions. With that in mind everything looks good! A new ticket will be created to ensure the logic is all correct and some of the previous comments should probably get moved into new ticket to help with QA.

randimays reviewed (approved):
LGTM! Great work!

thejordanwood commented:
Review screen looks good. Approved!
